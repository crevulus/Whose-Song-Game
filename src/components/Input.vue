<template>
  <div>
    <div class="mx-auto" style="max-width: 800px">
      <div class="w-full max-w-lg mx-auto">
        <div
          v-show="showAlert"
          class="bg-red-100 border border-red-400 text-red-700 px-4 py-3 rounded relative content-center"
          role="alert"
          style="max-width: 100%; margin: 20px auto;"
        >
          <strong class="font-bold">Too late!&nbsp;</strong>
          <span>The game has already started...</span>
          <span
            class="absolute top-0 bottom-0 right-0 px-4 py-3 rounded-r-sm bg-red-600"
            style="cursor: pointer"
          >
            <div class="text-white" @click="redirect">
              <title>Close</title>
              <strong>OK</strong>
            </div>
          </span>
        </div>
        <h4 class="title text-left font-semibold p-1">Search for your favourite song</h4>
        <t-input
          v-on:keyup="searchTrack"
          type="text"
          v-model="searchField"
          placeholder="Artists, tracks, or albums"
          class="w-full mb-4"
          maxlength="50"
        />
        <ul>
          <li
            class="flex bg-white p-3 rounded-md relative items-center border-transparent border-2 cursor-pointer"
            v-for="(song, idx) in songList"
            :class="{ 'border-purple-600': song.isSelected, 'hover:border-purple-200': !song.isSelected }"
            @click="$set(song, 'isSelected', !song.isSelected);selectSong(song, idx)"
            :key="idx"
          >
            <img class="rounded h-12 w-12" :src="song.image" />
            <div class="flex flex-col justify-center ml-4 pr-12">
              <p class="m-0" style="color:#667eea;">
                <strong>{{song.title}}</strong>
              </p>
              <p class="m-0">{{song.artists}}</p>
            </div>
            <div
              class="absolute right-0 top-0 h-full flex items-center"
              :class="{ hidden: !song.isSelected}"
            >
              <Checkbox />
            </div>
          </li>
        </ul>
        <div class="text-right">
          <t-button
            @click="confirm()"
            class="ml-auto w-48"
            variant="primary"
            :disabled="selectedSong.length === 0 || showAlert"
          >Submit song</t-button>
        </div>
      </div>
    </div>
  </div>
</template>

<script>
import axios from "axios";
import { mapMutations, mapGetters } from "vuex";
import { API, graphqlOperation } from "aws-amplify";
import commonMethods from "@/mixins/commonMethods";
import * as mutations from "@/graphql/mutations";
import * as subscriptions from "@/graphql/subscriptions";
import Checkbox from "@/components/Checkbox";

export default {
  name: "Input",
  components: { Checkbox },
  mixins: [commonMethods],
  data: function() {
    return {
      searchField: "",
      songList: [],
      selectedSong: "",
      timeoutId: null,
      showAlert: false
      // activityInstanceId: this.$route.params.activityInstanceId
    };
  },
  computed: {
    ...mapGetters(["accessToken"])
  },
  created() {
    this.updatedActivityInstanceSubscription();
    this.getActivityInstanceQuery();
    this.getAccessToken();
  },
  methods: {
    ...mapMutations(["setAccessToken"]),
    searchTrack(e) {
      const throttleSpeed = 400; // ms
      e.preventDefault();
      clearTimeout(this.timeoutId);

      this.timeoutId = setTimeout(() => {
        if (!this.searchField.length) return;

        const url = `https://api.spotify.com/v1/search?q=${this.searchField}&type=track,album,artist&limit=4`;
        axios
          .get(url, {
            headers: {
              Authorization: "Bearer " + this.accessToken
            }
          })
          .then(res => {
            this.songList = this.normalizeTrackData(res.data.tracks.items);
          });
      }, throttleSpeed);
    },
    normalizeTrackData(tracks) {
      return tracks.map(track => ({
        id: track.id,
        title: track.name,
        artists: track.artists.map(artist => artist.name).join(", "),
        album: track.album.name,
        image: track.album.images[2].url,
        isSelected: false
      }));
    },
    selectSong(song, idx) {
      this.songList.forEach((song, i) => (song.isSelected = idx === i));

      this.selectedSong = song;
    },
    redirect() {
      this.$router.push({
        name: "home"
      });
    },
    confirm() {
      if (this.status !== "waiting") {
        this.showAlert = true;
        return;
      }
      const { id, title, artists } = this.selectedSong;
      API.graphql(
        graphqlOperation(mutations.whoseSongUpdateActivityInstanceData, {
          activityInstanceId: this.activityInstanceId,
          userId: this.deviceId,
          action: "submitNewSong",
          trackId: id,
          trackTitle: title,
          trackArtists: artists
        })
      ).then(() => {
        this.$router.push({
          name: "lobby",
          params: {
            activityInstanceId: this.activityInstanceId
          }
        });
      });
    },
    updatedActivityInstanceSubscription() {
      API.graphql(
        graphqlOperation(subscriptions.updatedActivityInstance, {
          activityInstanceId: this.activityInstanceId
        })
      ).subscribe(response => {
        const data = response.value.data.updatedActivityInstance;
        this.status = data.status;
      });
    },
    getAccessToken() {
      const url = "https://accounts.spotify.com/api/token";
      const client_id = "f27c6cba06be4c7691eadadeb40bb8a8";
      const client_secret = "3ed8adb1529d4d35b4dd3abfcb1ec638";
      const hash = Buffer.from(client_id + ":" + client_secret).toString(
        "base64"
      );
      const config = {
        headers: {
          Authorization: "Basic " + hash,
          "Content-Type": "application/x-www-form-urlencoded"
        }
      };

      axios
        .post(url, "grant_type=client_credentials", config)
        .then(r => {
          const token = r.data.access_token;
          this.setAccessToken(token);
        })
        .catch(r => console.log(r));
    }
  }
};
</script>
