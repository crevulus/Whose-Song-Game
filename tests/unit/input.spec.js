import { mount, config, createLocalVue } from "@vue/test-utils";
import Input from "@/components/Input";
import VueTailwind from "vue-tailwind";

const localVue = createLocalVue();
localVue.use(VueTailwind);

config.mocks["$gtm"] = { trackView: function (name, path) { } };
config.mocks["$route"] = { params: { activityInstanceId: '123' } };

describe.only('SearchSong Component', () => {
  const commonMethods = {};

  const users = [
    {
      userId: '123',
      name: 'Foo'
    },
    {
      userId: '456',
      name: 'Bar'
    }
  ]

  const methods = {
    updatedActivityInstanceSubscription() {
      return
    },
    getActivityInstanceQuery() {
      return
    },
    getAccessToken() {
      return;
    }
  }

  const data = {
    activityInstanceId: '123456789',
    isHost: false,
    users,
    hostId: '456',
    status: 'started'
  }

  const tracks = [{
    id: 123,
    name: 'foo',
    artists: [{ name: 'Tell' }],
    album: { name: 'Mitch', images: { 2: { url: '123' } } }
  },
  {
    id: 123,
    name: 'foo',
    artists: [{ name: 'Tell' }],
    album: { name: 'Mitch', images: { 2: { url: '123' } } }
  }];

  const wrapper = mount(Input, {
    localVue,
    data() {
      return data;
    },
    methods,
    mixins: [commonMethods]
  })

  // checks if vue component
  test('is a Vue instance', () => {
    expect(wrapper.isVueInstance()).toBeTruthy()
  })

  // snapshot testing
  test('renders correctly', () => {
    expect(wrapper.element).toMatchSnapshot()
  })

  // calling methods
  it('testing normalized tracks', () => {
    const normalizedTracks = wrapper.vm.normalizeTrackData(tracks);
    expect(normalizedTracks).toEqual([{
      id: 123, title: 'foo', artists: 'Tell', album:
        'Mitch', "image": "123", "isSelected": false
    }, {
      id: 123, title: 'foo', artists: 'Tell', album:
        'Mitch', "image": "123", "isSelected": false
    }])
  })

  // async testing
  it('testing if page updates with songs', async () => {
    const normalizedTracks = wrapper.vm.normalizeTrackData(tracks);
    wrapper.setData({ songList: normalizedTracks });
    await wrapper.vm.$nextTick()
    const li = wrapper.find('li');
    expect(li.text()).toMatch('Tell')
  })
})






