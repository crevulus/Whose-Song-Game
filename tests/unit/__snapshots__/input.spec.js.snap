// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchSong Component renders correctly 1`] = `
<div
  class="mx-auto"
  style="max-width: 800px;"
>
  <div
    class="w-full max-w-lg mx-auto"
  >
    <h4
      class="title text-left font-semibold p-1"
    >
      Search for your favourite song
    </h4>
     
    <input
      class="w-full mb-4 t-input t-input-size-default t-input-status-default border block rounded p-3 bg-white"
      maxlength="50"
      placeholder="Artists, tracks, or albums"
      type="text"
    />
     
    <ul />
     
    <div
      class="text-right"
    >
      <button
        class="ml-auto w-48 t-button t-button-size-default border block rounded inline-flex items-center justify-center t-button-disabled cursor-not-allowed opacity-75 px-6 py-3 text-white bg-blue-500 border-blue-500 hover:bg-blue-600 hover:border-blue-600"
        disabled="disabled"
        type="button"
      >
        Submit song
      </button>
    </div>
  </div>
</div>
`;
